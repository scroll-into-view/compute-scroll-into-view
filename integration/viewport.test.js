beforeAll(async () => {
  await page.goto('http://localhost:3000/integration/viewport')
})

const ScrollLogicalPosition = ['start', 'center', 'end', 'nearest']

describe('scrollMode: always', () => {
  ScrollLogicalPosition.forEach(block => {
    ScrollLogicalPosition.forEach(inline => {
      test(`block: ${block}, inline: ${inline}`, async () => {
        const expected = await page.evaluate(
          options => {
            window.scrollTo(0, 0)
            document.querySelector('.target').scrollIntoView(options)
            const { scrollLeft, scrollTop } = document.scrollingElement
            return { left: scrollLeft, top: scrollTop }
          },
          { block, inline }
        )
        const actual = await page.evaluate(
          options => {
            window.scrollTo(0, 0)
            const [{ left, top }] = window.computeScrollIntoView(
              document.querySelector('.target'),
              options
            )
            return { left, top }
          },
          { block, inline }
        )
        expect(expected).toEqual(actual)
        expect(actual).toMatchSnapshot()

        // The 'nearest' cases can have diff behavior depending on the current scroll position
        if (block === 'nearest' || inline === 'nearest') {
          const expected = await page.evaluate(
            options => {
              window.scrollTo(window.innerWidth * 3, window.innerHeight * 3)
              document.querySelector('.target').scrollIntoView(options)
              const { scrollLeft, scrollTop } = document.scrollingElement
              return { left: scrollLeft, top: scrollTop }
            },
            { block, inline }
          )
          const actual = await page.evaluate(
            options => {
              window.scrollTo(window.innerWidth * 3, window.innerHeight * 3)
              const [{ left, top }] = window.computeScrollIntoView(
                document.querySelector('.target'),
                options
              )
              return { left, top }
            },
            { block, inline }
          )
          expect(expected).toEqual(actual)
          expect(actual).toMatchSnapshot()
        }
      })
    })
  })
})

describe('scrollMode: if-needed', () => {
  describe('vertical', () => {
    test('completely below the fold', async () => {
      const actual = await page.evaluate(() => {
        window.scrollTo(window.innerWidth * 1.5, window.innerHeight)
        return window.computeScrollIntoView(document.querySelector('.target'), {
          scrollMode: 'if-needed',
        })
      })
      expect(actual).toHaveLength(1)
      expect(actual).toMatchSnapshot()
    })

    test('partially below the fold', async () => {
      const actual = await page.evaluate(() => {
        window.scrollTo(window.innerWidth * 1.5, window.innerHeight + 50)
        return window.computeScrollIntoView(document.querySelector('.target'), {
          scrollMode: 'if-needed',
        })
      })
      expect(actual).toHaveLength(1)
      expect(actual).toMatchSnapshot()
    })
  })

  describe('horizontal', () => {
    test('completely overflowing', async () => {
      const actual = await page.evaluate(() => {
        window.scrollTo(window.innerWidth, window.innerHeight * 1.5)
        return window.computeScrollIntoView(document.querySelector('.target'), {
          scrollMode: 'if-needed',
        })
      })
      expect(actual).toHaveLength(1)
      expect(actual).toMatchSnapshot()
    })

    test('partially overflowing', async () => {
      const actual = await page.evaluate(() => {
        window.scrollTo(window.innerWidth + 50, window.innerHeight * 1.5)
        return window.computeScrollIntoView(document.querySelector('.target'), {
          scrollMode: 'if-needed',
        })
      })
      expect(actual).toHaveLength(1)
      expect(actual).toMatchSnapshot()
    })
  })
})
