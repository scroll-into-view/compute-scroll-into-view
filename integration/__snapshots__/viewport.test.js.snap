// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`scrollMode: always block: center, inline: center 1`] = `
Object {
  "left": 1250,
  "top": 950,
}
`;

exports[`scrollMode: always block: center, inline: end 1`] = `
Object {
  "left": 900,
  "top": 950,
}
`;

exports[`scrollMode: always block: center, inline: nearest 1`] = `
Object {
  "left": 900,
  "top": 950,
}
`;

exports[`scrollMode: always block: center, inline: nearest 2`] = `
Object {
  "left": 1600,
  "top": 950,
}
`;

exports[`scrollMode: always block: center, inline: start 1`] = `
Object {
  "left": 1600,
  "top": 950,
}
`;

exports[`scrollMode: always block: end, inline: center 1`] = `
Object {
  "left": 1250,
  "top": 700,
}
`;

exports[`scrollMode: always block: end, inline: end 1`] = `
Object {
  "left": 900,
  "top": 700,
}
`;

exports[`scrollMode: always block: end, inline: nearest 1`] = `
Object {
  "left": 900,
  "top": 700,
}
`;

exports[`scrollMode: always block: end, inline: nearest 2`] = `
Object {
  "left": 1600,
  "top": 700,
}
`;

exports[`scrollMode: always block: end, inline: start 1`] = `
Object {
  "left": 1600,
  "top": 700,
}
`;

exports[`scrollMode: always block: nearest, inline: center 1`] = `
Object {
  "left": 1250,
  "top": 700,
}
`;

exports[`scrollMode: always block: nearest, inline: center 2`] = `
Object {
  "left": 1250,
  "top": 1200,
}
`;

exports[`scrollMode: always block: nearest, inline: end 1`] = `
Object {
  "left": 900,
  "top": 700,
}
`;

exports[`scrollMode: always block: nearest, inline: end 2`] = `
Object {
  "left": 900,
  "top": 1200,
}
`;

exports[`scrollMode: always block: nearest, inline: nearest 1`] = `
Object {
  "left": 900,
  "top": 700,
}
`;

exports[`scrollMode: always block: nearest, inline: nearest 2`] = `
Object {
  "left": 1600,
  "top": 1200,
}
`;

exports[`scrollMode: always block: nearest, inline: start 1`] = `
Object {
  "left": 1600,
  "top": 700,
}
`;

exports[`scrollMode: always block: nearest, inline: start 2`] = `
Object {
  "left": 1600,
  "top": 1200,
}
`;

exports[`scrollMode: always block: start, inline: center 1`] = `
Object {
  "left": 1250,
  "top": 1200,
}
`;

exports[`scrollMode: always block: start, inline: end 1`] = `
Object {
  "left": 900,
  "top": 1200,
}
`;

exports[`scrollMode: always block: start, inline: nearest 1`] = `
Object {
  "left": 900,
  "top": 1200,
}
`;

exports[`scrollMode: always block: start, inline: nearest 2`] = `
Object {
  "left": 1600,
  "top": 1200,
}
`;

exports[`scrollMode: always block: start, inline: start 1`] = `
Object {
  "left": 1600,
  "top": 1200,
}
`;

exports[`scrollMode: if-needed horizontal completely overflowing 1`] = `
Array [
  Object {
    "el": "html",
    "left": 900,
    "top": 950,
  },
]
`;

exports[`scrollMode: if-needed horizontal fully negative overflowing 1`] = `
Array [
  Object {
    "el": "html",
    "left": 1600,
    "top": 950,
  },
]
`;

exports[`scrollMode: if-needed horizontal partially negative overflowing 1`] = `
Array [
  Object {
    "el": "html",
    "left": 1600,
    "top": 950,
  },
]
`;

exports[`scrollMode: if-needed horizontal partially overflowing 1`] = `
Array [
  Object {
    "el": "html",
    "left": 900,
    "top": 950,
  },
]
`;

exports[`scrollMode: if-needed vertical completely above the fold 1`] = `
Array [
  Object {
    "el": "html",
    "left": 1200,
    "top": 950,
  },
]
`;

exports[`scrollMode: if-needed vertical completely below the fold 1`] = `
Array [
  Object {
    "el": "html",
    "left": 1200,
    "top": 950,
  },
]
`;

exports[`scrollMode: if-needed vertical partially above the fold 1`] = `
Array [
  Object {
    "el": "html",
    "left": 1200,
    "top": 950,
  },
]
`;

exports[`scrollMode: if-needed vertical partially below the fold 1`] = `
Array [
  Object {
    "el": "html",
    "left": 1200,
    "top": 950,
  },
]
`;
